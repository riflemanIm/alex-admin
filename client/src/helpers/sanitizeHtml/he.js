/* @flow */
import invariant from "tiny-invariant";
/* eslint-disable no-control-regex */

// All BMP symbols that are not ASCII newlines, printable ASCII symbols, or
// code points listed in the first column of the overrides table on
// https://html.specOrName.whatwg.org/multipage/syntax.html#table-charref-overrides.
// const regexBmpWhitelist = /[\x01-\t\x0B\f\x0E-\x1F\x7F\x81\x8D\x8F\x90\x9D\xA0-\uFFFF]/g;
// All astral symbols.
const regexAstralSymbols = /[\uD800-\uDBFF][\uDC00-\uDFFF]/g;

const regexEscape = /["&'<>`]/g;

const escapeMap = {
  '"': "&quot;",
  "&": "&amp;",
  "'": "&#x27;",
  "<": "&lt;",
  // See https://mathiasbynens.be/notes/ambiguous-ampersands: in HTML, the
  // following is not strictly necessary unless it’s part of a tag or an
  // unquoted attribute value. We’re only escaping it to support those
  // situations, and for XML support.
  ">": "&gt;",
  // In Internet Explorer ≤ 8, the backtick character can be used
  // to break out of (un)quoted attribute values or HTML comments.
  // See http://html5sec.org/#102, http://html5sec.org/#108, and
  // http://html5sec.org/#133.
  "`": "&#x60;",
};

const regexInvalidEntity = /&#(?:[xX][^a-fA-F0-9]|[^0-9xX])/;

const regexInvalidRawCodePoint = /[\0-\x08\x0B\x0E-\x1F\x7F-\x9F\uFDD0-\uFDEF\uFFFE\uFFFF]|[\uD83F\uD87F\uD8BF\uD8FF\uD93F\uD97F\uD9BF\uD9FF\uDA3F\uDA7F\uDABF\uDAFF\uDB3F\uDB7F\uDBBF\uDBFF][\uDFFE\uDFFF]|[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF]/;

const regexDecode = /&#([0-9]+)(;?)|&#[xX]([a-fA-F0-9]+)(;?)|&([0-9a-zA-Z]+);|&(nbsp|quot|AMP|amp|GT|LT|gt|lt)([=a-zA-Z0-9])?/g;

const decodeMap = {
  amp: "&",
  AMP: "&",
  gt: ">",
  Gt: "\u226B",
  GT: ">",
  lt: "<",
  Lt: "\u226A",
  LT: "<",
  quot: '"',
  QUOT: '"',
  nbsp: "\xA0",
};

const decodeMapNumeric = {
  "0": "\uFFFD",
  "128": "\u20AC",
  "130": "\u201A",
  "131": "\u0192",
  "132": "\u201E",
  "133": "\u2026",
  "134": "\u2020",
  "135": "\u2021",
  "136": "\u02C6",
  "137": "\u2030",
  "138": "\u0160",
  "139": "\u2039",
  "140": "\u0152",
  "142": "\u017D",
  "145": "\u2018",
  "146": "\u2019",
  "147": "\u201C",
  "148": "\u201D",
  "149": "\u2022",
  "150": "\u2013",
  "151": "\u2014",
  "152": "\u02DC",
  "153": "\u2122",
  "154": "\u0161",
  "155": "\u203A",
  "156": "\u0153",
  "158": "\u017E",
  "159": "\u0178",
};

const invalidReferenceCodePoints = [
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  11,
  13,
  14,
  15,
  16,
  17,
  18,
  19,
  20,
  21,
  22,
  23,
  24,
  25,
  26,
  27,
  28,
  29,
  30,
  31,
  127,
  128,
  129,
  130,
  131,
  132,
  133,
  134,
  135,
  136,
  137,
  138,
  139,
  140,
  141,
  142,
  143,
  144,
  145,
  146,
  147,
  148,
  149,
  150,
  151,
  152,
  153,
  154,
  155,
  156,
  157,
  158,
  159,
  64976,
  64977,
  64978,
  64979,
  64980,
  64981,
  64982,
  64983,
  64984,
  64985,
  64986,
  64987,
  64988,
  64989,
  64990,
  64991,
  64992,
  64993,
  64994,
  64995,
  64996,
  64997,
  64998,
  64999,
  65000,
  65001,
  65002,
  65003,
  65004,
  65005,
  65006,
  65007,
  65534,
  65535,
  131070,
  131071,
  196606,
  196607,
  262142,
  262143,
  327678,
  327679,
  393214,
  393215,
  458750,
  458751,
  524286,
  524287,
  589822,
  589823,
  655358,
  655359,
  720894,
  720895,
  786430,
  786431,
  851966,
  851967,
  917502,
  917503,
  983038,
  983039,
  1048574,
  1048575,
  1114110,
  1114111,
];

/*--------------------------------------------------------------------------*/

const stringFromCharCode = String.fromCharCode;
const object = {};
const hasOwnProperty = object.hasOwnProperty;
const has = function (object, propertyName) {
  return hasOwnProperty.call(object, propertyName);
};

// Modified version of `ucs2encode`; see https://mths.be/punycode.
const codePointToSymbol = function (codePoint, original, errors) {
  let output = "";
  if ((codePoint >= 0xd800 && codePoint <= 0xdfff) || codePoint > 0x10ffff) {
    // See issue #4:
    // “Otherwise, if the number is in the range 0xD800 to 0xDFFF or is
    // greater than 0x10FFFF, then this is a parse error. Return a U+FFFD
    // REPLACEMENT CHARACTER.”

    parseError(
      `character reference "${original}" outside the permissible Unicode range`,
      errors
    );

    return "\uFFFD";
  }

  if (has(decodeMapNumeric, codePoint)) {
    parseError(`disallowed character reference ${original}`, errors);
    return decodeMapNumeric[codePoint];
  }

  if (invalidReferenceCodePoints.includes(codePoint)) {
    parseError(`disallowed character reference ${original}`, errors);
    // do not decode disallowed codes
    return original;
  }

  if (codePoint > 0xffff) {
    codePoint -= 0x10000;
    output += stringFromCharCode(((codePoint >>> 10) & 0x3ff) | 0xd800);
    codePoint = 0xdc00 | (codePoint & 0x3ff);
  }
  output += stringFromCharCode(codePoint);
  return output;
};

const hexEscape = function (codePoint) {
  return "&#x" + codePoint.toString(16).toUpperCase() + ";";
};

const parseError = function (message, errors) {
  // throw Error('Parse error: ' + message);
  if (!errors.includes(message)) {
    errors.push(message);
  }
};

/*
var escapeBmpSymbol = function(symbol) {
  return hexEscape(symbol.charCodeAt(0));
};
*/
/*--------------------------------------------------------------------------*/

export const encode = function (string) {
  const errors = [];

  if (regexInvalidRawCodePoint.test(string)) {
    parseError("forbidden code point", errors);
  }

  const escapeCodePoint = hexEscape;

  // Apply named character references.
  // Encode `<>"'&` using named character references.
  string = string.replace(regexEscape, function (string) {
    invariant(string in escapeMap, `escapeMap must contain ${string}`);

    return escapeMap[string]; // no need to check `has()` here
  });

  return {
    errors,
    encoded: string
      // Encode astral symbols.
      .replace(regexAstralSymbols, function ($0) {
        // https://mathiasbynens.be/notes/javascript-encoding#surrogate-formulae
        const high = $0.charCodeAt(0);
        const low = $0.charCodeAt(1);
        const codePoint = (high - 0xd800) * 0x400 + low - 0xdc00 + 0x10000;
        return escapeCodePoint(codePoint);
      }),
    // .replace(regexBmpWhitelist, escapeBmpSymbol),
  };
};
// Expose default options (so they can be overridden globally).

export const decode = function (html) {
  const errors = [];

  if (regexInvalidEntity.test(html)) {
    parseError("malformed character reference", errors);
  }

  return {
    errors,
    decoded: html.replace(regexDecode, ($0, $1, $2, $3, $4, $5, $6) => {
      let codePoint;
      let semicolon;
      let decDigits;
      let hexDigits;
      let reference;

      if ($1) {
        // Decode decimal escapes, e.g. `&#119558;`.
        decDigits = $1;
        semicolon = $2;
        if (!semicolon) {
          parseError(
            `character reference ${$0} was not terminated by a semicolon`,
            errors
          );
        }
        // eslint-disable-next-line
        codePoint = parseInt(decDigits, 10);
        return codePointToSymbol(codePoint, $0, errors);
      }
      if ($3) {
        // Decode hexadecimal escapes, e.g. `&#x1D306;`.
        hexDigits = $3;
        semicolon = $4;
        if (!semicolon) {
          parseError(
            `character reference ${$0} was not terminated by a semicolon`,
            errors
          );
        }
        // eslint-disable-next-line
        codePoint = parseInt(hexDigits, 16);
        return codePointToSymbol(codePoint, $0, errors);
      }
      if ($5) {
        // Decode named character references with trailing `;`, e.g. `&copy;`.
        reference = $5;
        if (has(decodeMap, reference)) {
          return decodeMap[reference];
        } else {
          // Ambiguous ampersand. https://mths.be/notes/ambiguous-ampersands
          parseError(
            `ambiguous ampersand or unknown named character ${$0}`,
            errors
          );
          return $0;
        }
      }
      // If we’re still here, it’s a legacy reference for sure. No need for an
      // extra `if` check.
      // Decode named character references without trailing `;`, e.g. `&amp`
      // This is only a parse error if it gets converted to `&`, or if it is
      // followed by `=` in an attribute context.
      reference = $6;

      parseError(
        `named character reference ${$0} was not terminated by a semicolon`,
        errors
      );
      // Note: there is no need to check `has(decodeMapLegacy, reference)`.
      return $0 + ";";
    }),
  };
};
