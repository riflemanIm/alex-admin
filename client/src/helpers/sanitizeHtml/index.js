/* @flow */

export { sanitizeHtmlWithErrors, sanitizeHtml } from './sanitizeHtml';
