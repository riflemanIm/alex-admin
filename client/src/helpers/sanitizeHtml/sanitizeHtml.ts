/* @flow */

import { encode, decode } from "./he";

const tagRe = /<([/])?([\S]+?)((?:\s+[^\s=>/]+(?:\s*=\s*(?:(?:"[^"]*")|(?:'[^']*')|[^>\s/]+))?)*?)\s*(\/?)\s*>/gim;
const tagComments = /<!--[\s\S]*?-->/gim;

var rattrs = /([^\s=]+)(?:\s*=\s*(?:(?:"((?:[^"])*)")|(?:'((?:[^'])*)')|([^>\s]+)))?/gim;

const cleanAttributeName = (name: string) => {
  // attribute can have any non tab, form-feed name, but we
  // use more strict check it can be \w:-
  var re = /[^\w\d:-]+/gim;
  return name.replace(re, "").toLowerCase();
};

export interface Options {
  tagsBlacklist: ReadonlyArray<string | ((string: string) => boolean)>;
  attrsBlacklist: ReadonlyArray<string | ((string: string) => boolean)>;
  attrsWithScheme: ReadonlyArray<string | ((string: string) => boolean)>;
  allowedSchemes: ReadonlyArray<string | ((string: string) => boolean)>;
  selfClosingTags: ReadonlyArray<string | ((string: string) => boolean)>;
}

const defaultOptions = {
  tagsBlacklist: [
    "script",
    "style",
    "iframe",
    "object",
    "meta",
    (tagName: string) => {
      const reTag = /^[a-zA-Z]+\w*$/i;
      return !reTag.test(tagName);
    },
  ],

  attrsBlacklist: [
    "poster",
    "repeat",
    "pattern",
    "srcset",
    (attrName: string) => attrName.startsWith("on"),
  ],

  attrsWithScheme: ["formaction", "href", "xlink:href"],
  allowedSchemes: [
    "http",
    "https",
    "mailto",
    "tel",
    (attrValue: string) => !attrValue.includes(":"),
  ],
  selfClosingTags: ["br"],
};

const isSchemaAllowed = (
  value: string,
  allowedSchemes: ReadonlyArray<string | ((string: string) => boolean)>
): boolean =>
  allowedSchemes.some((schemePrefixOrTestFn) => {
    if (typeof schemePrefixOrTestFn === "string") {
      return value.startsWith(`${schemePrefixOrTestFn}:`);
    }
    return schemePrefixOrTestFn(value);
  });

const isListHasItem = (
  name: string,
  list: ReadonlyArray<string | ((string: string) => boolean)>
) =>
  list.some((itemNameOrTestFn) => {
    if (typeof itemNameOrTestFn === "string") {
      return itemNameOrTestFn === name;
    }
    return itemNameOrTestFn(name);
  });

export const sanitizeHtmlWithErrors = (
  str: string,
  argOptions?: Options
): {
  errors: ReadonlyArray<string>;
  value: string;
} => {
  let errors: any[] = [];
  const stack: any[] = [];
  const options = {
    ...defaultOptions,
    ...argOptions,
  };

  const replaceTag = (
    match: string,
    closeSym: string,
    tagName: string,
    attrs: string,
    selfCloseSym: string
  ) => {
    if (isListHasItem(tagName, options.tagsBlacklist)) {
      return "";
    }

    if (closeSym === "/") {
      // fix html if closed tags don't match
      const idx = stack.lastIndexOf(tagName);
      let r = "";
      if (idx > -1) {
        for (let i = stack.length - 1; i >= idx; --i) {
          r += `</${stack[i]}>`;
          stack.pop();
        }
        return r;
      }

      return "";
    }

    rattrs.lastIndex = 0;

    let result: RegExpExecArray | null;
    let attrsArr = [];
    while ((result = rattrs.exec(attrs))) {
      const [
        ,
        attrName,
        attrQuoteValue,
        attrSingleQuoteValue,
        attrNoQuoteValue,
      ] = result;

      if (attrName == null) continue;

      const cleanName = cleanAttributeName(attrName);

      if (isListHasItem(attrName, options.attrsBlacklist)) {
        continue;
      }

      if (cleanName === "") continue;

      const { decoded, errors: errorsDec } = decode(
        attrQuoteValue || attrSingleQuoteValue || attrNoQuoteValue || ""
      );

      const { encoded, errors: errorsEnc } = encode(decoded);

      errors = errors.concat(errorsDec, errorsEnc);

      if (
        isListHasItem(attrName, options.attrsWithScheme) &&
        !isSchemaAllowed(encoded, options.allowedSchemes)
      ) {
        continue;
      }

      attrsArr.push(encoded ? `${cleanName}="${encoded}"` : `${cleanName}`);
    }

    if (
      selfCloseSym !== "/" &&
      options.selfClosingTags.every((v) => v !== tagName.toLowerCase())
    ) {
      stack.push(tagName);
    }

    const newAttrs = attrsArr.join(" ");
    return newAttrs === ""
      ? selfCloseSym === "/" &&
        options.selfClosingTags.some((v) => v === tagName.toLowerCase())
        ? `<${tagName}/>`
        : `<${tagName}>${selfCloseSym === "/" ? `</${tagName}>` : ""}`
      : `<${tagName} ${newAttrs}>${
          selfCloseSym === "/" ? `</${tagName}>` : ""
        }`;
  };

  let r = str.replace(tagComments, "").replace(tagRe, replaceTag);
  for (let i = stack.length - 1; i > -1; --i) {
    r += `</${stack[i]}>`;
  }

  return {
    errors,
    value: r,
  };
};

export const sanitizeHtml = (str: string, argOptions?: Options): string => {
  const { value } = sanitizeHtmlWithErrors(str, argOptions);
  return value;
};
