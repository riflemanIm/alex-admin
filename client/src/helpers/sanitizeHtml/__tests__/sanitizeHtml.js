/* @flow */

import { sanitizeHtml } from '../sanitizeHtml';

test('it restores html closing tags', () => {
  expect(
    sanitizeHtml(`
    <b><a><a><strong>hello</b></a>
    <b><a><a><strong>hello</b></a>
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeHtml(`
    <b><a><a><strong>hello
  `)
  ).toMatchSnapshot();
});

test('it works with broken tags', () => {
  expect(
    sanitizeHtml(`
    <a <script>hello world
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeHtml(`
    <a>hello world</b</a>
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeHtml(`
    <e<<a <a>hello world</b</a>
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeHtml(`
    <e <<a <a>hello world</b</a>
  `)
  ).toMatchSnapshot();
});

test('it works with gt lt symbols inside tags', () => {
  expect(
    sanitizeHtml(`
    <a>hello > world</a>
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeHtml(`
    <a>hello>world</a>
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeHtml(`
    <a>hello < world</a>
  `)
  ).toMatchSnapshot();

  // in this case <word is ambigious so will be stripped out
  // same will do chrome
  expect(
    sanitizeHtml(`
    <a>hello<world</a>
  `)
  ).toMatchSnapshot();
});

test('it strips script style iframe object meta tags', () => {
  expect(
    sanitizeHtml(`
    <script language=js>script</script>
    <script>script</script>
    <meta>meta</meta>
    <style>style</style>
    <iframe>iframe</iframe>
    <object>object</object>
  `)
  ).toMatchSnapshot();
});

test('it strips tags not matched with [a-zA-Z]+\\w* (default options)', () => {
  expect(
    sanitizeHtml(`
    <1a>hello world</1a>
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeHtml(`
    <плохойТег><good>hello world</good></плохойТег>
  `)
  ).toMatchSnapshot();
});

test('it strips poster,repeat,pattern,srcset,on* attributes', () => {
  expect(
    sanitizeHtml(`
    <test poster="hi all" yes="no" repeat="all hi" no="yes" pattern="">hello world</test>
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeHtml(`
    <test srcset="hi all" yes="no" onhello="all hi" no="yes" onopen="">hello world</test>
  `)
  ).toMatchSnapshot();
});

test('it preserves formaction,href,xlink:href attributes with schemas http,https,mailto,tel,no schema', () => {
  expect(
    sanitizeHtml(`
    <a href="tel:hello/world">hello</a>
  `)
  ).toMatchSnapshot();
  expect(
    sanitizeHtml(`
    <a href="mailto:hello/world">hello</a>
  `)
  ).toMatchSnapshot();
  expect(
    sanitizeHtml(`
    <a href="/hello/world">hello</a>
  `)
  ).toMatchSnapshot();
  expect(
    sanitizeHtml(`
    <a href="http://hello/world">hello</a>
  `)
  ).toMatchSnapshot();
  expect(
    sanitizeHtml(`
    <a href="https://hello/world">hello</a>
  `)
  ).toMatchSnapshot();
});

test('it strips formaction,href,xlink:href attributes with schemas other than http,https,mailto,tel,no schema', () => {
  expect(
    sanitizeHtml(`
    <a href="hello:hello/world">hello</a>
  `)
  ).toMatchSnapshot();
  expect(
    sanitizeHtml(`
    <a href="javascript:hello/world">hello</a>
  `)
  ).toMatchSnapshot();
  expect(
    sanitizeHtml(`
    <a xlink:href="javascript:hello/world">hello</a>
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeHtml(`
    <button formaction="javascript:hello/world">hello</button>
  `)
  ).toMatchSnapshot();
});

test('it understand \' in rounded with " in attributes', () => {
  expect(
    sanitizeHtml(`
    <a test="hey'hey">hello</a>
  `)
  ).toMatchSnapshot();
});

test('it understand " in rounded with \' in attributes', () => {
  expect(
    sanitizeHtml(`
    <a test='hey"hey'>hello</a>
  `)
  ).toMatchSnapshot();
});

test('it not break tags because of < > in attributes', () => {
  expect(
    sanitizeHtml(`
    <a test='hey<hey'>hello</a>
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeHtml(`
    <a test='hey>hey'>hello</a>
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeHtml(`
    <a test='<hey>hey</hey>'>hello</a>
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeHtml(`
    <a test="hey<hey">hello</a>
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeHtml(`
    <a test="hey>hey">hello</a>
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeHtml(`
    <a test="<hey>hey</hey>">hello</a>
  `)
  ).toMatchSnapshot();
});

test('it understand self closing tags', () => {
  expect(
    sanitizeHtml(`
    <b/><a test="lala" />
  `)
  ).toMatchSnapshot();
});

test('it not encodes non ascii chars in attributes', () => {
  expect(
    sanitizeHtml(`
    <a test="привет мир" />
  `)
  ).toMatchSnapshot();
});

test('it encodes \'"><& in attributes', () => {
  expect(
    sanitizeHtml(`
    <a test="'&><" />
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeHtml(`
    <a test=a&b />
  `)
  ).toMatchSnapshot();
});

test('it fixes non wrapped with quotes attributes', () => {
  expect(
    sanitizeHtml(`
    <a test=hello />
  `)
  ).toMatchSnapshot();
});

test('it can get wrong schema even if it encoded', () => {
  expect(
    sanitizeHtml(`
    <a href="&#106;&#97;&#118;&#97;&#115;&#99;&#114;&#105;&#112;&#116;&#58;&#97;&#108;&#101;&#114;&#116;&#40;&#39;&#88;&#83;&#83;&#39;&#41;">Hax</a>
  `)
  ).toMatchSnapshot();
  expect(
    sanitizeHtml(`
    <a href="&#106&#97&#118&#97&#115&#99&#114&#105&#112&#116&#58&#97&#108&#101&#114&#116&#40&#39&#88&#83&#83&#39;&#41;">Hax</a>
  `)
  ).toMatchSnapshot();
});

test('it can get right schema even if it encoded', () => {
  expect(
    sanitizeHtml(`
    <a href="&#116;&#101;&#108;&#58;&#104;&#101;&#108;&#108;&#111;">Hax</a>
  `)
  ).toMatchSnapshot();

  expect(
    sanitizeHtml(`
    <a href="&#116&#101&#108&#58&#104&#101&#108&#108&#111">Hax</a>
  `)
  ).toMatchSnapshot();
});

test('something bad will somehow be sanitized', () => {
  expect(
    sanitizeHtml(`
    <a <script a=">" b=">">ddd</a>
  <div hi="игрушек>игрушечники">ddd</div>
  <div мир игрушек>игрушечники</div>
  <жопа мира />
  <script

  href="hello">world</script>
  <b><a href='jopa' /></b>
  <div/>
  >c><b
    hi='al"ll lol'
    let=me&not
  >hello<a>popopopo</b></a>

  <!-- жопа мира -->
  <a xlink:href="hello" onjaja="ddd" srcset="jajaj" href="&#106;&#97;&#118;&#97;&#115;&#99;&#114;&#105;&#112;&#116;&#58;&#97;&#108;&#101;&#114;&#116;&#40;&#39;&#88;&#83;&#83;&#39;&#41;">Hax</a>
  <a href="javascript:hiall">ddd</a>
  `)
  ).toMatchSnapshot();
});
