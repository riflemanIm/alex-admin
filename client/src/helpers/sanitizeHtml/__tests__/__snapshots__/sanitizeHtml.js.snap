// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it can get right schema even if it encoded 1`] = `
"
    <a href=\\"tel:hello\\">Hax</a>
  "
`;

exports[`it can get right schema even if it encoded 2`] = `
"
    <a href=\\"tel:hello\\">Hax</a>
  "
`;

exports[`it can get wrong schema even if it encoded 1`] = `
"
    <a>Hax</a>
  "
`;

exports[`it can get wrong schema even if it encoded 2`] = `
"
    <a>Hax</a>
  "
`;

exports[`it encodes '"><& in attributes 1`] = `
"
    <a test=\\"&#x27;&amp;&gt;&lt;\\"></a>
  "
`;

exports[`it encodes '"><& in attributes 2`] = `
"
    <a test=\\"a&amp;b\\"></a>
  "
`;

exports[`it fixes non wrapped with quotes attributes 1`] = `
"
    <a test=\\"hello\\"></a>
  "
`;

exports[`it not break tags because of < > in attributes 1`] = `
"
    <a test=\\"hey&lt;hey\\">hello</a>
  "
`;

exports[`it not break tags because of < > in attributes 2`] = `
"
    <a test=\\"hey&gt;hey\\">hello</a>
  "
`;

exports[`it not break tags because of < > in attributes 3`] = `
"
    <a test=\\"&lt;hey&gt;hey&lt;/hey&gt;\\">hello</a>
  "
`;

exports[`it not break tags because of < > in attributes 4`] = `
"
    <a test=\\"hey&lt;hey\\">hello</a>
  "
`;

exports[`it not break tags because of < > in attributes 5`] = `
"
    <a test=\\"hey&gt;hey\\">hello</a>
  "
`;

exports[`it not break tags because of < > in attributes 6`] = `
"
    <a test=\\"&lt;hey&gt;hey&lt;/hey&gt;\\">hello</a>
  "
`;

exports[`it not encodes non ascii chars in attributes 1`] = `
"
    <a test=\\"привет мир\\"></a>
  "
`;

exports[`it preserves formaction,href,xlink:href attributes with schemas http,https,mailto,tel,no schema 1`] = `
"
    <a href=\\"tel:hello/world\\">hello</a>
  "
`;

exports[`it preserves formaction,href,xlink:href attributes with schemas http,https,mailto,tel,no schema 2`] = `
"
    <a href=\\"mailto:hello/world\\">hello</a>
  "
`;

exports[`it preserves formaction,href,xlink:href attributes with schemas http,https,mailto,tel,no schema 3`] = `
"
    <a href=\\"/hello/world\\">hello</a>
  "
`;

exports[`it preserves formaction,href,xlink:href attributes with schemas http,https,mailto,tel,no schema 4`] = `
"
    <a href=\\"http://hello/world\\">hello</a>
  "
`;

exports[`it preserves formaction,href,xlink:href attributes with schemas http,https,mailto,tel,no schema 5`] = `
"
    <a href=\\"https://hello/world\\">hello</a>
  "
`;

exports[`it restores html closing tags 1`] = `
"
    <b><a><a><strong>hello</strong></a></a></b>
    <b><a><a><strong>hello</strong></a></a></b>
  "
`;

exports[`it restores html closing tags 2`] = `
"
    <b><a><a><strong>hello
  </strong></a></a></b>"
`;

exports[`it strips formaction,href,xlink:href attributes with schemas other than http,https,mailto,tel,no schema 1`] = `
"
    <a>hello</a>
  "
`;

exports[`it strips formaction,href,xlink:href attributes with schemas other than http,https,mailto,tel,no schema 2`] = `
"
    <a>hello</a>
  "
`;

exports[`it strips formaction,href,xlink:href attributes with schemas other than http,https,mailto,tel,no schema 3`] = `
"
    <a>hello</a>
  "
`;

exports[`it strips formaction,href,xlink:href attributes with schemas other than http,https,mailto,tel,no schema 4`] = `
"
    <button>hello</button>
  "
`;

exports[`it strips poster,repeat,pattern,srcset,on* attributes 1`] = `
"
    <test yes=\\"no\\" no=\\"yes\\">hello world</test>
  "
`;

exports[`it strips poster,repeat,pattern,srcset,on* attributes 2`] = `
"
    <test yes=\\"no\\" no=\\"yes\\">hello world</test>
  "
`;

exports[`it strips script style iframe object meta tags 1`] = `
"
    script
    script
    meta
    style
    iframe
    object
  "
`;

exports[`it strips tags not matched with [a-zA-Z]+\\w* (default options) 1`] = `
"
    hello world
  "
`;

exports[`it strips tags not matched with [a-zA-Z]+\\w* (default options) 2`] = `
"
    <good>hello world</good>
  "
`;

exports[`it understand " in rounded with ' in attributes 1`] = `
"
    <a test=\\"hey&quot;hey\\">hello</a>
  "
`;

exports[`it understand ' in rounded with " in attributes 1`] = `
"
    <a test=\\"hey&#x27;hey\\">hello</a>
  "
`;

exports[`it understand self closing tags 1`] = `
"
    <b></b><a test=\\"lala\\"></a>
  "
`;

exports[`it works with broken tags 1`] = `
"
    <a script>hello world
  </a>"
`;

exports[`it works with broken tags 2`] = `
"
    <a>hello world
  </a>"
`;

exports[`it works with broken tags 3`] = `
"
    hello world
  "
`;

exports[`it works with broken tags 4`] = `
"
    <e a a>hello world
  </e>"
`;

exports[`it works with gt lt symbols inside tags 1`] = `
"
    <a>hello > world</a>
  "
`;

exports[`it works with gt lt symbols inside tags 2`] = `
"
    <a>hello>world</a>
  "
`;

exports[`it works with gt lt symbols inside tags 3`] = `
"
    <a>hello < world</a>
  "
`;

exports[`it works with gt lt symbols inside tags 4`] = `
"
    <a>hello
  </a>"
`;

exports[`something bad will somehow be sanitized 1`] = `
"
    <a script a=\\"&gt;\\" b=\\"&gt;\\">ddd</a>
  <div hi=\\"игрушек&gt;игрушечники\\">ddd</div>
  <div>игрушечники</div>
  
  world
  <b><a href=\\"jopa\\"></a></b>
  <div></div>
  >c><b hi=\\"al&quot;ll lol\\" let=\\"me&amp;not\\">hello<a>popopopo</a></b>

  
  <a xlink:href=\\"hello\\">Hax</a>
  <a>ddd</a>
  "
`;
